pkg/oci/authn/client.go

```diff
@@ -31,100 +30,109 @@ const (
 
 // Options used for the HTTP client that can authenticate with auth.Credentials or via OAuth2.0 Options Credentials flow.
 type Options struct {
-	Ctx               context.Context
-	Credentials       *auth.Credential
-	Oauth             bool
-	ClientCredentials *clientcredentials.Config
+	Ctx                   context.Context
+	CredentialsFuncsCache map[string]func(context.Context, string) (auth.Credential, error)
+	CredentialsFuncs      []func(context.Context, string) (auth.Credential, error)
+	AutoLoginHandler      *AutoLoginHandler
 }
 
 // NewClient creates a new authenticated client to interact with a remote registry.
-func NewClient(options ...func(*Options)) remote.Client {
-	opt := &Options{}
+func NewClient(options ...func(*Options)) *auth.Client {
+	opt := &Options{
+		CredentialsFuncsCache: make(map[string]func(context.Context, string) (auth.Credential, error)),
+	}
 
 	for _, o := range options {
 		o(opt)
 	}
 
-	if opt.Oauth && opt.ClientCredentials != nil {
-		return opt.ClientCredentials.Client(opt.Ctx)
-	} else {
-		authClient := auth.Client{
-			Client: &http.Client{
-				Transport: &http.Transport{
-					Proxy: http.ProxyFromEnvironment,
-					DialContext: (&net.Dialer{
-						Timeout:   30 * time.Second,
-						KeepAlive: 30 * time.Second,
-					}).DialContext,
-					ForceAttemptHTTP2:     true,
-					MaxIdleConns:          100,
-					IdleConnTimeout:       90 * time.Second,
-					TLSHandshakeTimeout:   10 * time.Second,
-					ExpectContinueTimeout: 1 * time.Second,
-					// TODO(loresuso, alacuku): tls config.
-				},
-			},
-			Cache: auth.NewCache(),
-			Credential: func(ctx context.Context, registry string) (auth.Credential, error) {
-				return *opt.Credentials, nil
+	authClient := auth.Client{
+		Client: &http.Client{
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				DialContext: (&net.Dialer{
+					Timeout:   30 * time.Second,
+					KeepAlive: 30 * time.Second,
+				}).DialContext,
+				ForceAttemptHTTP2:     true,
+				MaxIdleConns:          100,
+				IdleConnTimeout:       90 * time.Second,
+				TLSHandshakeTimeout:   10 * time.Second,
+				ExpectContinueTimeout: 1 * time.Second,
+				// TODO(loresuso, alacuku): tls config.
 			},
-		}
+		},
+		Cache: auth.NewCache(),
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			// try cred func from cache first
+			credFunc, exists := opt.CredentialsFuncsCache[reg]
+			if exists {
+				return credFunc(ctx, reg)
+			}
+
+			// if auto login is on check if we tried logging in to registry
+			if opt.AutoLoginHandler != nil {
```

Comment:
This should work fine since the flow is

1. Do autologin
2. Retrieve credentials

so autologin always happens before trying to use the credentials.

Reply:
Looking at the code, the flow is: 
 1. The store is loaded into memory when the client is created;
 2. Do autologin, and store the new credential;
 3. Retrieve credentials from the store previously loaded. At this point, we only have the credentials before the autologin happened.

Reply:
Okay yeah upon re-examining it i see the problem. I had a slight misconception in how the CredentialStore works here.
The flow would work for any CredentialStore other than the `FileStore` since for that one it seems to read the information from the loaded config file. For other types of CredentialStores its a bit more dynamic so they retrieve the info upon `Get` call which is where my misconception is from.

The above flow can still work if we ensure the client uses the same `CredentialStore` instance in both the auto login and the basic auth credential function.

Reply:
If you are okay with it i'd go with the one CredentialStore instance per Client option. The global singelton variant seems like the less elegant solution.

Reply:
I've updated the code with that solution in mind for now.

Reply:
I just saw the new code and It triggers a stackoverflow. The `client` allocates a new `AutoLoginHandler` that allocates a new `client` and so on. We need to break the circular dependency somehow.

Reply:
Anyway, I agree to share the store between the client/credentialFuncs and the autoLoginHandler.

Reply:
Ok fixed the recursion.

Reply:
i see that a test seems to be failing in CI/CD now 🤔 
>   [FAILED] Got stuck at:
      ERRO: unable to connect to remote registry "localhost:45267": Get "http://localhost:45267/v2/": dial tcp [::1]:45267: connect: connection refused
 >   
 > Waiting for:
      ERRO: expecting 1 rulesfile object received 2: invalid number of rulesfiles
    
   but i cant seem to reproduce the test error locally even after rebasing to latest main branch

Reply:
Yeah I double checked and the test that failed in CI/CD runs fine. I can't really figure out why that test might have failed in the pipeline, but runs fine locally. Only thing i could imagine would be that the [registry startup in the BeforeSuite](https://github.com/falcosecurity/falcoctl/blob/be3cbd69bff41caacf6105c04d512f016372bec1/pkg/oci/pusher/pusher_suite_test.go#L60-L63) has a race condition.

I tested a bit by adding `time.Sleep(time.Millisecond * 10)` in front of it and indeed it will make the tests fail sometimes with the same connection error. Not sure how to fix this race condition since starting the registry seems to be a blocking call.

Reply:
Yeah, there is a race condition. The fix would be to wait for the container registry to be up before starting the test suites. 

I will fix it in the coming days.